File: experiments-local/experiment-15/runs/agent=omp_language=go_model=openrouter/deepseek/deepseek-v3.2_tooling=none/rep1/internal/db/book_repository.go
package db

import (
	"database/sql"

	"bookapi/internal/models"
)

func GetAllBooks(authorFilter string) ([]models.Book, error) {
	var rows *sql.Rows
	var err error

	if authorFilter != "" {
		rows, err = DB.Query("SELECT id, title, author, year, isbn FROM books WHERE author LIKE ?", "%"+authorFilter+"%")
	} else {
		rows, err = DB.Query("SELECT id, title, author, year, isbn FROM books")
	}

	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var books []models.Book
	for rows.Next() {
		var book models.Book
		err := rows.Scan(&book.ID, &book.Title, &book.Author, &book.Year, &book.ISBN)
		if err != nil {
			return nil, err
		}
		books = append(books, book)
	}

	return books, nil
}

func GetBookByID(id int) (*models.Book, error) {
	var book models.Book
	err := DB.QueryRow("SELECT id, title, author, year, isbn FROM books WHERE id = ?", id).
		Scan(&book.ID, &book.Title, &book.Author, &book.Year, &book.ISBN)

	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}

	return &book, nil
}

func CreateBook(book *models.Book) error {
	result, err := DB.Exec("INSERT INTO books (title, author, year, isbn) VALUES (?, ?, ?, ?)",
		book.Title, book.Author, book.Year, book.ISBN)
	if err != nil {
		return err
	}

	id, err := result.LastInsertId()
	if err != nil {
		return err
	}
	book.ID = int(id)
	return nil
}

func UpdateBook(id int, book *models.Book) error {
	result, err := DB.Exec("UPDATE books SET title = ?, author = ?, year = ?, isbn = ? WHERE id = ?",
		book.Title, book.Author, book.Year, book.ISBN, id)
	if err != nil {
		return err
	}

	rowsAffected, err := result.RowsAffected()
	if err != nil {
		return err
	}
	if rowsAffected == 0 {
		return sql.ErrNoRows
	}

	return nil
}

func DeleteBook(id int) error {
	result, err := DB.Exec("DELETE FROM books WHERE id = ?", id)
	if err != nil {
		return err
	}

	rowsAffected, err := result.RowsAffected()
	if err != nil {
		return err
	}
	if rowsAffected == 0 {
		return sql.ErrNoRows
	}

	return nil
}

File: experiments-local/experiment-15/runs/agent=omp_language=java_model=openrouter/qwen/qwen3.7-plus_tooling=none/rep2/src/main/java/com/example/bookapi/BookRepository.java
package com.example.bookapi;

import org.springframework.dao.EmptyResultDataAccessException;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.RowMapper;
import org.springframework.jdbc.support.GeneratedKeyHolder;
import org.springframework.jdbc.support.KeyHolder;
import org.springframework.stereotype.Repository;

import java.sql.PreparedStatement;
import java.sql.Statement;
import java.sql.Types;
import java.util.List;
import java.util.Optional;

@Repository
public class BookRepository {
    private final JdbcTemplate jdbcTemplate;

    public BookRepository(JdbcTemplate jdbcTemplate) {
        this.jdbcTemplate = jdbcTemplate;
    }

    private final RowMapper<Book> rowMapper = (rs, rowNum) -> new Book(
            rs.getLong("id"),
            rs.getString("title"),
            rs.getString("author"),
            rs.getObject("year", Integer.class),
            rs.getString("isbn")
    );

    public Book save(Book book) {
        String sql = "INSERT INTO books (title, author, year, isbn) VALUES (?, ?, ?, ?)";
        KeyHolder keyHolder = new GeneratedKeyHolder();
        jdbcTemplate.update(connection -> {
            PreparedStatement ps = connection.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS);
            ps.setString(1, book.title());
            ps.setString(2, book.author());
            if (book.year() != null) {
                ps.setInt(3, book.year());
            } else {
                ps.setNull(3, Types.INTEGER);
            }
            ps.setString(4, book.isbn());
            return ps;
        }, keyHolder);
        
        Long id = keyHolder.getKey().longValue();
        return new Book(id, book.title(), book.author(), book.year(), book.isbn());
    }

    public List<Book> findAll(String authorFilter) {
        if (authorFilter != null && !authorFilter.trim().isEmpty()) {
            return jdbcTemplate.query("SELECT * FROM books WHERE author LIKE ?", rowMapper, "%" + authorFilter + "%");
        }
        return jdbcTemplate.query("SELECT * FROM books", rowMapper);
    }

    public Optional<Book> findById(Long id) {
        try {
            Book book = jdbcTemplate.queryForObject("SELECT * FROM books WHERE id = ?", rowMapper, id);
            return Optional.of(book);
        } catch (EmptyResultDataAccessException e) {
            return Optional.empty();
        }
    }

    public boolean update(Book book) {
        String sql = "UPDATE books SET title = ?, author = ?, year = ?, isbn = ? WHERE id = ?";
        int rows = jdbcTemplate.update(sql, book.title(), book.author(), book.year(), book.isbn(), book.id());
        return rows > 0;
    }

    public boolean deleteById(Long id) {
        int rows = jdbcTemplate.update("DELETE FROM books WHERE id = ?", id);
        return rows > 0;
    }
}